### PR TITLE
feat[venom]: improve liveness computation

### PR DESCRIPTION
**Pull request by charles-cooper**
### What I did
move dfs order calculation from domtree to cfg analysis.

remove unnecessary calculation of domtree in sccp

remove redundant `IRFunction.compute_reachability`

change cfg_out order

refactor shared phi fixup code

### Commit message

```
feat[venom]: improve liveness computation

traversing in reverse topsort order improves stack scheduling slightly

this commit also adds a topsort method to CFGAnalysis, and speeds it up
by only checking the terminator instruction instead of iterating over
all the instructions in every basic block.

additional refactors:
- move dfs order calculation from domtree to cfg analysis.
- remove unnecessary calculation of domtree in sccp
- remove redundant IRFunction.compute_reachability
- change cfg_out order
- refactor shared phi fixup code
- remove useless `__eq__()` and `__hash__()` for IRBasicBlock
```

### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

File: vyper/venom/venom_to_assembly.py [hunk 1]
```diff
@@ -308,7 +314,7 @@ def _generate_evm_for_basicblock_r(
 
         ref.extend(asm)
 
-        for bb in basicblock.reachable:
+        for bb in basicblock.cfg_out:
```
Comment (HodanPlodky) [inline vyper/venom/venom_to_assembly.py:317]: Could be reachable removed from basic block if it is no longer used here?

Comment (charles-cooper) [reply vyper/venom/venom_to_assembly.py:317]: Yes, good point

Comment (charles-cooper) [reply vyper/venom/venom_to_assembly.py:317]: https://github.com/vyperlang/vyper/pull/4330/commits/1b117c8e9f51701501fe0ba277d063d4bf4a4a69

**Comment by charles-cooper** [comment]
hmm, this creates a regression in `branch_storm.vy`

**Review by harkal — APPROVED** [review]

**Pull request merged**